ethcore/sync/src/chain/supplier.rs

```diff
@@ -145,14 +145,47 @@ impl SyncSupplier {
 				}
 			};
 
-			result.unwrap_or_else(|e| {
-				debug!(target:"sync", "{} -> Malformed packet {} : {}", peer, packet_id, e);
-			})
+			match result {
+				Err(PacketProcessError::Decoder(e)) => debug!(target:"sync", "{} -> Malformed packet {} : {}", peer, packet_id, e),
+				Err(PacketProcessError::ClientBusy) => sync.write().add_delayed_request(peer, packet_id, data),
```

Comment:
Maybe we can ensure that only `GetBlockHeadersPacket`s get added here instead of performing that check when processing delayed requests?

Reply:
I've tried to create a generic mechanism of packets with delayed response. IMO it's pretty useful and may be used in future